ion-schema/src/main/kotlin/com/amazon/ionschema/internal/DeferredReference.kt

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.ionschema.internal
+
+import com.amazon.ion.IonSymbol
+import com.amazon.ion.IonValue
+import com.amazon.ionschema.InvalidSchemaException
+import com.amazon.ionschema.Violations
+
+/**
+ * Represents a [TypeInternal] that is known to be declared in a schema document, but the [TypeInternal] instance has
+ * not necessarily been created yet. A [DeferredReference] can be resolved or unresolved. When it is unresolved,
+ * it is unknown whether the referred [TypeInternal] instance exists. When it is resolved, this [DeferredReference]
+ * is able to return and/or delegate validation to the referred [TypeInternal] instance.
```

Comment:
(Suggestion) Reading this it seems like we will always get a `TypeInternal`  but that might not be true in the cases where it is unsolvable. I think when it is resolved it has 2 possibilities either get `TypeInternal` or null.
Do you think we should add that in the comment here?

```suggestion
 * it is unknown whether the referred [TypeInternal] instance exists. When it is resolved, this [DeferredReference]
 * is able to return and/or delegate validation to the referred [TypeInternal] instance and if it not resolvable then it will return `null`. 
```

Reply:
What I am trying to say is that this class represents a `type::{ name: foo, ... }` that _is_ known to exist. However, the `TypeInternal` might not be instantiated yet. If it is not instantiated yet, then we cannot resolve this `DeferredReference`. Calling `resolve()` should only be done by the `DeferredReferenceManager` so that we can be sure it happens only at the correct time.

I will try to refactor these classes to better encapsulate the behavior, and I will fix the documentation.

Reply:
That would be great. Thanks.